### PR TITLE
Add option "Block merge if PR needs work".

### LINKED ISSUE #56: Any way to block automerge if there is "Needs work" from one of the reviewers?
Imagine we have 6 developers in the team and 4 reviewers are required to merge (auto or manually) the PR.

If one of the reviewers sets "Needs work" on a PR, automerge still happens if 4 reviewers will approve it and CI will pass.

Is there any way to prevent auto merge if there is at least one "Needs work" on a PR? Thanks!

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Closes #56.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -33,6 +33,7 @@
       <groupId>com.atlassian.bitbucket.server</groupId>
       <artifactId>bitbucket-api</artifactId>
       <scope>provided</scope>
+      <version>4.11.1</version>
```
Comment (arturdryomov) [inline pom.xml:36]: Is it really necessary?

File: src/main/java/com/monitorjbl/plugins/MergeBlocker.java [hunk 2]
```diff
@@ -37,8 +39,26 @@ public void check(@Nonnull MergeRequest mergeRequest) {
...
+        final Boolean blockAutoMergeBecausePrNeedsWork = config.getBlockMergeIfPrNeedsWork() && needsWork(pr);
+
+        if (blockAutoMergeBecausePrNeedsWork) {
+          mergeRequest.veto("PR marked as Needs Work from reviewer(s)", "Resolve it before merge.");
+        }
       }
     }
   }
 
+  private boolean needsWork(final PullRequest pr) {
```
Comment (arturdryomov) [inline src/main/java/com/monitorjbl/plugins/MergeBlocker.java:52]: `hasNeedsWork`?

File: src/main/java/com/monitorjbl/plugins/MergeBlocker.java [hunk 3]
```diff
@@ -37,8 +39,26 @@ public void check(@Nonnull MergeRequest mergeRequest) {
...
+    boolean needsWork = false;
+
+    for (PullRequestParticipant reviewer : pr.getParticipants()) {
+      if (reviewer.getStatus() == PullRequestParticipantStatus.NEEDS_WORK) {
+        needsWork = true;
```
Comment (arturdryomov) [inline src/main/java/com/monitorjbl/plugins/MergeBlocker.java:57]: Just `return true`.

File: src/main/java/com/monitorjbl/plugins/MergeBlocker.java [hunk 4]
```diff
@@ -37,8 +39,26 @@ public void check(@Nonnull MergeRequest mergeRequest) {
...
+        break;
+      }
+    }
+
+    return needsWork;
```
Comment (arturdryomov) [inline src/main/java/com/monitorjbl/plugins/MergeBlocker.java:62]: Just `return false`.

File: src/main/java/com/monitorjbl/plugins/config/Config.java [hunk 5]
```diff
@@ -15,6 +15,7 @@
   private List<String> requiredReviewers = newArrayList();
   private List<String> requiredReviewerGroups = newArrayList();
   private Integer requiredReviews = 0;
+  private Boolean blockMergeIfPrNeedsWork = false;
```
Comment (arturdryomov) [inline src/main/java/com/monitorjbl/plugins/config/Config.java:18]: `blockMergeWhenHasNeedsWork`, `blockMergeIfHasNeedsWork`?

**Review by arturdryomov — APPROVED** [review]
Looks great!

File: pom.xml (hunk 1 above)
Comment (artem-zinnatullin) [reply pom.xml:36]: Yes, otherwise I couldn't use `NEEDS_WORK` enum item

File: src/main/java/com/monitorjbl/plugins/MergeBlocker.java (hunk 2 above)
Comment (artem-zinnatullin) [reply src/main/java/com/monitorjbl/plugins/MergeBlocker.java:52]: Well, current version is perfectly aligned with Bitbucket UI: `Pull request needs work`.

File: src/main/java/com/monitorjbl/plugins/MergeBlocker.java (hunk 3 above)
Comment (artem-zinnatullin) [reply src/main/java/com/monitorjbl/plugins/MergeBlocker.java:57]: lol, k

File: src/main/java/com/monitorjbl/plugins/config/Config.java (hunk 5 above)
Comment (artem-zinnatullin) [reply src/main/java/com/monitorjbl/plugins/config/Config.java:18]: "Pull Request Needs Work", I don't think it should be changed.

File: src/test/java/com/monitorjbl/plugins/MergeBlockerTest.java [hunk 6]
```diff
@@ -153,4 +154,58 @@ public void testBlocking_reviewerIsAuthor_approved() throws Exception {
     sut.check(merge);
     verify(merge, never()).veto(anyString(), anyString());
   }
+
+  @Test
+  public void testBlocking_prNeedsWorkEnabledAndOneNeedsWorkSet() throws Exception {
```
Comment (monitorjbl) [inline src/test/java/com/monitorjbl/plugins/MergeBlockerTest.java:159]: Yay, test cases! 👍

**Comment by monitorjbl** [comment]
Looks good to me, thanks for the PR!

**Comment by artem-zinnatullin** [comment]
YYYYayay!

**Pull request merged**